source/extensions/filters/udp/dns_filter/dns_filter.cc

```
@@ -3,55 +3,59 @@
 
 namespace Envoy {
 namespace Extensions {
 namespace UdpFilters {
 namespace DnsFilter {
 
+static constexpr std::chrono::milliseconds DEFAULT_RESOLVER_TIMEOUT{500};
+static constexpr std::chrono::seconds DEFAULT_RESOLVER_TTL{300};
+
 DnsFilterEnvoyConfig::DnsFilterEnvoyConfig(
     Server::Configuration::ListenerFactoryContext& context,
     const envoy::extensions::filters::udp::dns_filter::v3alpha::DnsFilterConfig& config)
-    : root_scope_(context.scope()), stats_(generateStats(config.stat_prefix(), root_scope_)) {
+    : root_scope_(context.scope()), cluster_manager_(context.clusterManager()), api_(context.api()),
+      stats_(generateStats(config.stat_prefix(), root_scope_)) {
   using envoy::extensions::filters::udp::dns_filter::v3alpha::DnsFilterConfig;
 
-  static constexpr std::chrono::milliseconds DEFAULT_RESOLVER_TIMEOUT{500};
-  static constexpr std::chrono::seconds DEFAULT_RESOLVER_TTL{300};
-
   const auto& server_config = config.server_config();
 
-  // TODO(abaptiste): Read the external DataSource
-  if (server_config.has_inline_dns_table()) {
-    const auto& dns_table = server_config.inline_dns_table();
-    const size_t entries = dns_table.virtual_domains().size();
-
-    virtual_domains_.reserve(entries);
-    for (const auto& virtual_domain : dns_table.virtual_domains()) {
-      AddressConstPtrVec addrs{};
-      if (virtual_domain.endpoint().has_address_list()) {
-        const auto& address_list = virtual_domain.endpoint().address_list().address();
-        addrs.reserve(address_list.size());
-        // This will throw an exception if the configured_address string is malformed
-        for (const auto& configured_address : address_list) {
-          const auto ipaddr =
-              Network::Utility::parseInternetAddress(configured_address, 0 /* port */);
-          addrs.push_back(ipaddr);
-        }
+  envoy::data::dns::v3::DnsTable dns_table;
+  bool result = loadServerConfig(server_config, dns_table);
```

question: How does this failure get signaled to the control-plane? Like if a user supplies bad config from an xDS endpoint. Dont we need to throw an exception or something? Not sure the correct API for it...

---

I would recommend asking Marco. He implemented something like that for the lambda filter, where Envoy would just NACK a bad configuration.

---

I'll take a look and see what I can find.

---

The 'convention' seems to be that we throw an EnvoyException.   The code I used to load the data source was mirrored from elsewhere in Envoy, complete with the exception catching.

The filter does work even if the data source isn't read.  There are messages in the logs that will indicate a failure.

---

> The filter does work even if the data source isn't read

Ah I see.